File: hackrank-sep-math/huge-donation.cpp
#include <iostream>
#include <string>
#include <cmath>

using namespace std;

typedef unsigned long long int LL;

inline LL compute(LL n) {
	return (n * (n+1)/2) * (2*n+1) / 3;
}

LL binarySearch(LL lo, LL hi, const LL& x) {
	while (lo < hi) {
		LL mid = lo + (hi - lo + 1) / 2;
		LL sum = compute(mid);
		if (sum == x) return mid;
		if (sum > x) hi = mid - 1;
		else lo = mid;
	}
	return lo;
}

int main() {
	int t; cin >> t;
	while (t--) {
		LL x; cin >> x;
		cout << binarySearch(1, sqrtl(x), x) << endl;
	}	

	return 0;
}

File: hackrank-sep-math/huge-donation.py
from math import sqrt

def compute(n):
	return n * (n+1) * (2*n+1)/6

def search(lo, hi, x):
	while lo < hi:
		mid = lo + (hi - lo + 1) / 2
		sum = compute(mid)
		if sum == x:
			return mid
		if sum > x:
			hi = mid - 1
		else:
			lo = mid
	return lo

if __name__ == "__main__":
	t = int(raw_input())
	for _ in xrange(t):
		x = int(raw_input())
		print search(1, int(sqrt(x)), x)